man/man1/sqlr-import.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.14.
.TH SQLR-IMPORT "1" "March 2021" "SQL Relay" "User Commands"
.SH NAME
sqlr-import \- manual page for sqlr-import 1.9.0
.SH SYNOPSIS
.B sqlr-import
[\fI\,OPTIONS\/\fR]
.SH DESCRIPTION
sqlr\-import is the SQL Relay database object import utility.
.PP
Import a database object from a file created previously or elsewhere using
sqlr\-export.
.SH OPTIONS
.SS "Connection options:"
.TP
\fB\-host\fR host
Host name or IP address of the server to
connect to.
.TP
\fB\-port\fR port
Port to connect to.
.TP
\fB\-socket\fR socket
Local unix socket file name to connect to.
Can be used instead of host/port for making
connections to local servers.
.TP
\fB\-user\fR user
User name to auth with.
.TP
\fB\-password\fR password
Password to auth with.
.SS "Alternate connection options:"
.TP
\fB\-config\fR config
Override the default configuration with the
specified configuration.
.TP
\fB\-id\fR instanceid
Derive connection info and credentials from the
specified instance, as defined in the
configuration.
.TP
\fB\-krb\fR
Use Kerberos authentication and encryption.
.TP
\fB\-krbservice\fR svc
Use the specified kerberos service.
.TP
\fB\-krbmech\fR mech
Use the specified kerberos mechanism.
.TP
\fB\-krbflags\fR flags
Use the specified kerberos flags,
comma\-separated.
.TP
\fB\-tls\fR
Use TLS/SSL authentication and encrpyiton.
.TP
\fB\-tlsversion\fR version
Use the specified TLS/SSL version.
.TP
\fB\-tlscert\fR file
Use the specified certificate chain file.
This file should contain the client's
certificate, private key, and signing
certificates, as appropriate.
On Windows systems, this must be a .pfx file.
On non\-Windows systems, a variety of file
formats are supported.
.TP
\fB\-tlspassword\fR pwd
Use the specified password to acess the private
key in the file specified by \fB\-tlscert\fR.
.TP
\fB\-tlsciphers\fR "list"
Allow the specified list of ciphers.  The
list should be quoted and the ciphers should be
separated by spaces.
.TP
\fB\-tlsvalidate\fR (no|ca|ca+host|ca+domain)
Certificate validation option.
"no" \- Don't validate the server's certificate.
"ca" \- Validate that the server's certificate
was signed by a trusted certificate authority.
"ca+host" \- Perform "ca" validation and also
validate that one of the subject alternate names
(or common name if no SANs are present) in the
certificate matches the host parameter.
(Falls back to "ca" validation when a unix
socket is used.)
"ca+domain" \- Perform "ca" validation and also
validate that the domain name of one of the
subject alternate naames (or common name if no
SANs are present) in the certificate matches
the domain name of the host parameter.
(Falls back to "ca" validation when a unix
socket is used.)
.TP
\fB\-tlsca\fR file
Use the specified certificate authority file
when validating the server's certificate.  Or,
if "file" is a directory, then use all
certificate authority files found in that
directory when validating the server's
certifictate.
.TP
\fB\-tlsdepth\fR depth
Set the maximum certificate chain validation
depth to the specified depth.
.SS "Command options:"
.TP
\fB\-file\fR filename
Import the specified file.
.TP
\fB\-commitcount\fR rowcount
Execute a commit each time the specified number
of rows have been imported since the beginning
or since the previous commit.
.TP
\fB\-debug\fR [filename]
Write debug information to the specified file
or to the screen if no file is specified.
.TP
\fB\-verbose\fR
Display details about the import process.
.SH EXAMPLES
Import a table and sequence using the server at svr:9000 as usr/pwd.
.IP
sqlr\-import \-host svr \-port 9000 \-user usr \-password pwd \e
.IP
\-file mytable.tbl
.IP
sqlr\-import \-host svr \-port 9000 \-user usr \-password pwd \e
.IP
\-file myseq.seq
.PP
Import a table and sequence using the local server on socket /tmp/svr.sock
as usr/pwd.
.IP
sqlr\-import \-socket /tmp/svr.sock \-user usr \-password pwd \e
.IP
\-file mytable.tbl
.IP
sqlr\-import \-socket /tmp/svr.sock \-user usr \-password pwd \e
.IP
\-file myseq.seq
.PP
Import a table and sequence using connection info and credentials from
an instance defined in the default configuration.
.IP
sqlr\-import \-id myinst \-file mytable.tbl
.IP
sqlr\-import \-id myinst \-file myseq.seq
.PP
Import a table and sequence using connection info and credentials from
in instance defined in the config file ./myconfig.conf
.IP
sqlr\-import \-config ./myconfig.conf \-id myinst \-file mytable.tbl
.IP
sqlr\-import \-config ./myconfig.conf \-id myinst \-file myseq.seq
.PP
Rudiments version: 1.4.0
Compiled: Mar  4 2021 02:32:28
.SH AUTHOR
Written by David Muse.
.SH COPYRIGHT
Copyright \(co 1999\-2018 David Muse
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
SQL Relay is a persistent database connection pooling, proxying, throttling, load balancing and query routing/filtering system for Unix and Linux supporting ODBC, Oracle, MySQL, PostgreSQL, Sybase, MS SQL Server, IBM DB2, Firebird, SQLite and MS Access (minimally) with APIs for C, C++, .NET, Perl, Perl-DBI, Python, Python-DB, PHP, PHP PDO, Ruby, Ruby-DBI, Java, TCL and Erlang, ODBC and ADO.NET drivers, drop-in replacement libraries for MySQL and PostgreSQL, command line clients and extensive documentation.  The APIs support advanced database operations such as substitution and bind variables, multi-row fetches, client-side result set caching and suspended transactions.  It is ideal for speeding up database-driven web-based applications, accessing databases from unsupported platforms, migrating between databases, distributing access to replicated or clustered databases and throttling database access.


src/cmdline/sqlr-import.cpp
// Copyright (c) 1999-2018 David Muse
// See the file COPYING for more information

#include <sqlrelay/sqlrclient.h>
#include <sqlrelay/sqlrimportxml.h>
#include <sqlrelay/sqlrimportcsv.h>
#include <sqlrelay/sqlrutil.h>
#include <rudiments/process.h>
#include <rudiments/stdio.h>
#include <config.h>
#include <defaults.h>
#include <version.h>

static void helpmessage(const char *progname) {
	stdoutput.printf(
		"%s is the %s database object import utility.\n"
		"\n"
		"Import a database object from a file created previously or elsewhere using\n"
		"%s-export.\n"
		"\n"
		"Usage: %s [OPTIONS]\n"
		"\n"
		"Options:\n"
		"\n"
		CONNECTIONOPTIONS
		"\n"
		"Command options:\n"
		"	-file filename		Import the specified file.\n"
		"\n"
		"	-commitcount rowcount	Execute a commit each time the specified number\n"
		"				of rows have been imported since the beginning\n"
		"				or since the previous commit.\n"
		"\n"
		"	-debug [filename]	Write debug information to the specified file\n"
		"				or to the screen if no file is specified.\n"
		"\n"
		"	-verbose		Display details about the import process.\n"
		"\n"
		"Examples:\n"
		"\n"
		"Import a table and sequence using the server at svr:9000 as usr/pwd.\n"
		"\n"
		"	%s -host svr -port 9000 -user usr -password pwd \\\n"
		"		-file mytable.tbl\n"
		"\n"
		"	%s -host svr -port 9000 -user usr -password pwd \\\n"
		"		-file myseq.seq\n"
		"\n"
		"Import a table and sequence using the local server on socket /tmp/svr.sock\n"
		"as usr/pwd.\n"
		"\n"
		"	%s -socket /tmp/svr.sock -user usr -password pwd \\\n"
		"		-file mytable.tbl\n"
		"\n"
		"	%s -socket /tmp/svr.sock -user usr -password pwd \\\n"
		"		-file myseq.seq\n"
		"\n"
		"Import a table and sequence using connection info and credentials from\n"
		"an instance defined in the default configuration.\n"
		"\n"
		"	%s -id myinst -file mytable.tbl\n"
		"\n"
		"	%s -id myinst -file myseq.seq\n"
		"\n"
		"Import a table and sequence using connection info and credentials from\n"
		"in instance defined in the config file ./myconfig.conf\n"
		"\n"
		"	%s -config ./myconfig.conf -id myinst -file mytable.tbl\n"
		"\n"
		"	%s -config ./myconfig.conf -id myinst -file myseq.seq\n"
		"\n",
		progname,SQL_RELAY,SQLR,progname,
		progname,progname,progname,progname,
		progname,progname,progname,progname);
}

int main(int argc, const char **argv) {

	version(argc,argv);
	help(argc,argv);

	sqlrcmdline 	cmdline(argc,argv);
	sqlrpaths	sqlrpth(&cmdline);
	sqlrconfigs	sqlrcfgs(&sqlrpth);

	// get command-line options
	const char	*configurl=sqlrpth.getConfigUrl();
	const char	*id=cmdline.getValue("id");
	const char	*host=cmdline.getValue("host");
	uint16_t	port=charstring::toInteger(
				(cmdline.found("port"))?
				cmdline.getValue("port"):DEFAULT_PORT);
	const char	*socket=cmdline.getValue("socket");
	const char	*user=cmdline.getValue("user");
	const char	*password=cmdline.getValue("password");
	bool		usekrb=cmdline.found("krb");
	const char	*krbservice=cmdline.getValue("krbservice");
	const char	*krbmech=cmdline.getValue("krbmech");
	const char	*krbflags=cmdline.getValue("krbflags");
	bool		usetls=cmdline.found("tls");
	const char	*tlscert=cmdline.getValue("tlscert");
	const char	*tlsversion=cmdline.getValue("tlsversion");
	const char	*tlspassword=cmdline.getValue("tlspassword");
	const char	*tlsciphers=cmdline.getValue("tlsciphers");
	const char	*tlsvalidate="no";
	if (cmdline.found("tlsvalidate")) {
		tlsvalidate=cmdline.getValue("tlsvalidate");
	}
	const char	*tlsca=cmdline.getValue("tlsca");
	uint16_t	tlsdepth=charstring::toUnsignedInteger(
					cmdline.getValue("tlsdepth"));
	const char	*file=cmdline.getValue("file");
	const char	*commitcountstr=cmdline.getValue("commitcount");
	bool		debug=cmdline.found("debug");
	const char	*debugfile=NULL;
	if (debug) {
		debugfile=cmdline.getValue("debug");
	}
	bool		verbose=cmdline.found("verbose");
	const char	*table=cmdline.getValue("table");
	bool 		ignorecolumns=cmdline.found("ignorecolumns");
	const char	*primarykeyname=
				cmdline.getValue("primarykeyname");
	uint32_t	primarykeyposition=
				charstring::toUnsignedInteger(
					cmdline.getValue("primarykeyposition"));
	const char	*primarykeysequence=
				cmdline.getValue("primarykeysequence");

	// at least id, host or socket, and file are required
	if ((charstring::isNullOrEmpty(id) &&
		charstring::isNullOrEmpty(host) &&
		charstring::isNullOrEmpty(socket)) ||
		charstring::isNullOrEmpty(file)) {

		stdoutput.printf("usage: \n"
			" %s-import -host host -port port -socket socket\n"
			"        [-user user -password password]\n"
			"        [-krb] [-krbservice svc] [-krbmech mech] "
			"[-krbflags flags]\n"
			"        [-tls] [-tlsversion version]\n"
			"        [-tlscert certfile] [-tlspassword password]\n"
			"        [-tlsciphers cipherlist]\n"
			"        [-tlsvalidate (yes|no)] [-tlsca ca] "
			"[-tlsdepth depth]\n"
			"        -file file [-commitcount rowcount]\n"
			"        [-debug [filename]] [-verbose]\n"
			"  or\n"
			" %s-import [-config config] -id id\n"
			"        -file file [-commitcount rowcount]\n"
			"        [-debug [filename]] [-verbose]\n",
			SQLR,SQLR);
		process::exit(1);
	}

	// if an id was specified, then get various values from the config file
	if (!charstring::isNullOrEmpty(id)) {
		sqlrconfig	*cfg=sqlrcfgs.load(configurl,id);
		if (cfg) {
			if (!cmdline.found("host")) {
				host="localhost";
			}
			if (!cmdline.found("port")) {
				port=cfg->getDefaultPort();
			}
			if (!cmdline.found("socket")) {
				socket=cfg->getDefaultSocket();
			}
			if (!cmdline.found("krb")) {
				usekrb=cfg->getDefaultKrb();
			}
			if (!cmdline.found("krbservice")) {
				krbservice=cfg->getDefaultKrbService();
			}
			if (!cmdline.found("krbmech")) {
				krbmech=cfg->getDefaultKrbMech();
			}
			if (!cmdline.found("krbflags")) {
				krbflags=cfg->getDefaultKrbFlags();
			}
			if (!cmdline.found("tls")) {
				usetls=cfg->getDefaultTls();
			}
			if (!cmdline.getValue("tlsciphers")) {
				tlsciphers=cfg->getDefaultTlsCiphers();
			}
			if (!cmdline.found("user")) {
				user=cfg->getDefaultUser();
				password=cfg->getDefaultPassword();
			}
		}
	}

	// configure sql relay connection
	sqlrconnection	sqlrcon(host,port,socket,user,password,0,1);
	sqlrcursor	sqlrcur(&sqlrcon);

	// configure kerberos
	if (usekrb) {
		sqlrcon.enableKerberos(krbservice,krbmech,krbflags);
	} else if (usetls) {
		sqlrcon.enableTls(tlsversion,tlscert,tlspassword,tlsciphers,
						tlsvalidate,tlsca,tlsdepth);
	}

	// configure debug
	if (debug) {
		if (debugfile) {
			sqlrcon.setDebugFile(debugfile);
		}
		sqlrcon.debugOn();
	}

	// set commitcount
	// (if not specified at all, default to 100)
	uint64_t	commitcount=
			(charstring::isNullOrEmpty(commitcountstr))?
				100:charstring::toInteger(commitcountstr);

	// set up logging
	logger	lg;
	stdoutdestination	std;
	lg.addLogDestination(&std);

	// xml or csv
	sqlrimport	*sqlri;
	if (!charstring::compareIgnoringCase(
			charstring::findLast(file,'.'),".csv")) {
		sqlri=new sqlrimportcsv();
		if (!charstring::isNullOrEmpty(primarykeyname) &&
			!charstring::isNullOrEmpty(primarykeysequence)) {
			((sqlrimportcsv *)sqlri)->insertPrimaryKey(
							primarykeyname,
							primarykeyposition,
							primarykeysequence);
		}
	} else {
		sqlri=new sqlrimportxml();
	}
	sqlri->setSqlrConnection(&sqlrcon);
	sqlri->setSqlrCursor(&sqlrcur);
	sqlri->setDbType(sqlrcon.identify());
	if (!charstring::isNullOrEmpty(table)) {
		sqlri->setObjectName(table);
	}
	sqlri->setIgnoreColumns(ignorecolumns);
	sqlri->setCommitCount(commitcount);
	if (verbose) {
		sqlri->setLogger(&lg);
		sqlri->setCoarseLogLevel(1);
		sqlri->setFineLogLevel(1);
		sqlri->setLogIndent(1);
	}
	bool	success=sqlri->importFromFile(file);
	delete sqlri;
	process::exit(!success);
}
